### DISCUSSION TITLE
Example on how to read rigid body bin files

### DISCUSSION THREAD

**Discussion by arjun-mani** [InteractiveComputerGraphics/SPlisHSPlasH#305] [Q&A] [answered]
Hello,

Thank you for this wonderful simulator. I used it copiously in our work "SurfsUp: Learning Fluid Simulation from Novel Surfaces" (https://surfsup.cs.columbia.edu/), published at ICCV 2023. I'm very grateful for your help with Github issues I raised, and for building and maintaining Splishsplash. Thanks very much.

I had a quick question about reading rigid body files exported during a simulation. I'm trying to read the .bin files that are exported, to get the rotation and translation attributes of the rigid body. However, I'm not able to read the bin files in Python. I can try to read the VTK files, but I think it is harder to extract the rotation and translation from those. I've realized some people have asked about the .bin files in the past. Could you provide a minimal Python example of how one could read these .bin files? That would be super helpful if possible.

If there's another simple way to get the rotation and translation of the rigid body during the simulation, that would also work. I'd prefer not to write my own exporter, unless it has to come to doing that. Would greatly appreciate any help. Thanks very much.

**Comment by janbender** [comment, marked as answer]
Hi, 

Nice Work!

The .bin format is very simple and I just introduced it to avoid exporting lots of mesh files. The first file contains information which do not change over time and the following files only contain the translation and rotation matrix for each rigid body. I wrote a Maya Python script to import these files in Maya. You find the relevant function here in line 1244:

https://github.com/InteractiveComputerGraphics/SPlisHSPlasH/blob/9696086df53ea75168314b3936f5d1665f38cbc3/Scripts/Maya/SPlisHSPlasH.py#L1244

The script loads the whole sequence. First, first frame file is loaded (line 1248). It contains the number of bodies (1254) and then for each body a string containing the name of the mesh OBJ file (1260-1263), the scale (1281-1283), a boolean "isWall" (1287), the color (1288-1291). (I use the isWall flag to make walls invisible since fluids are often occluded by walls.)

Then for all bodies: the translation (1306-1308) and the rotation (1311-1314).

All following files contain only translation and rotation for all bodies (1324-1348).

I would recommend to copy the code and to remove all Maya commands. 

Hope that helps!

**Reply by arjun-mani** [reply]
Hi, 

Thanks a lot for providing this code, I really appreciate it. I'm noticing some issues when I use the code -- I'm getting numBodies=0, and then the output of other fields is unreasonable (e.g. empty object file name, scaling of 0.0). Do you have an idea what might be causing the issue? I have two rigid bodies in the simulation, and VTK files are correctly being written out for each.

For reference, I am using Splishsplash 2.11.0. I saw something about fixing exports in 2.13.0, so I can try the new version. Not sure if that will fix the issue. Would appreciate any insight you may have on this matter. Thanks very much.

**Reply by janbender** [reply]
Hm, not sure without testing, but it could be a problem of float vs. double. Is USE_DOUBLE_PRECISION active in CMake?

**Reply by arjun-mani** [reply]
I don't think it is active. I can try reinstalling with this option.

**Reply by janbender** [reply]
Does the problem already occur in the first file? Since the number of bodies is the first data in the file, something different would be wrong in this case.

**Reply by arjun-mani** [reply]
Sorry, my mistake, it works fine! I didn't realize that the first file alone contained this information -- so I had grabbed all the struct.unpack lines but was processing the files in random order and applying the code. With this change it all works and I am able to read successfully. Apologies for not seeing that earlier. Thanks very much for your time and suggestions which were very helpful.

**Reply by arjun-mani** [reply]
Just one other quick question: I wanted to confirm that the rotation vector has the nine numbers of the matrix as [first row, second row, third row]? I guess this is termed row-order. Thanks.

**Reply by janbender** [reply]
The matrix is in row order.

**Comment by janbender** [comment]
Here are some example files which I generated with the simulator without USE_DOUBLE_PRECISION. I can load the files without problems with the script:
[rigid_bodies.zip](https://github.com/InteractiveComputerGraphics/SPlisHSPlasH/files/14184654/rigid_bodies.zip)
